File: src/main/kotlin/com/OfflinePlay/domain/notification/entity/Notification.kt
package com.contenido.domain.notification.entity

import com.contenido.domain.user.entity.User
import jakarta.persistence.*
import org.springframework.data.annotation.CreatedDate
import org.springframework.data.jpa.domain.support.AuditingEntityListener
import java.time.LocalDateTime

enum class NotificationType {
    NEW_EVENT, NEW_POST, NEW_COMMENT, NEW_LIKE,
    /** 크리에이터(기획자) 신청 결과 — 어드민이 승인/거절. */
    APPLICATION_APPROVED, APPLICATION_REJECTED,
    /** 이벤트 참가 신청이 접수됨 — 채널 owner(기획자)에게 발송. */
    PARTICIPATION_REQUESTED,
    /** 이벤트 참가 신청 결과 — 채널 owner(기획자)가 승인/거절. */
    PARTICIPATION_APPROVED, PARTICIPATION_REJECTED,
    /** APPROVED 상태에서 참가자가 직접 취소 — 채널 owner(기획자)에게 발송. */
    PARTICIPATION_CANCELED,
    /** 티켓 발급 완료 — 승인된 참가자(buyer)에게 발송. */
    TICKET_ISSUED,
    /** 현장 체크인 완료 — 참가자(buyer)에게 발송. */
    TICKET_CHECKED_IN,
}

@Entity
@Table(name = "notifications")
@EntityListeners(AuditingEntityListener::class)
class Notification(

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "receiver_id", nullable = false)
    val receiver: User,

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 30)
    val type: NotificationType,

    @Column(nullable = false)
    val title: String,

    @Column(nullable = false)
    val message: String,

    @Column(name = "target_type", nullable = false, length = 20)
    val targetType: String,

    @Column(name = "target_id", nullable = false)
    val targetId: Long,

    @Column(name = "is_read", nullable = false)
    var isRead: Boolean = false,
) {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long = 0

    @CreatedDate
    @Column(name = "created_at", nullable = false, updatable = false)
    lateinit var createdAt: LocalDateTime
        protected set

    fun markAsRead() {
        isRead = true
    }
}

File: frontend/src/types/notification.ts
export interface Notification {
  id: number
  type: string
  title: string
  message: string
  targetType: string
  targetId: number
  isRead: boolean
  createdAt: string
}

/**
 * PR95 — backend `NotificationType` enum 과 1:1. 추후 enum 이 추가되면 여기도 함께 갱신해
 * preferences UI 가 누락 없이 노출하도록 한다.
 */
export type NotificationType =
  | 'NEW_EVENT'
  | 'NEW_POST'
  | 'NEW_COMMENT'
  | 'NEW_LIKE'
  | 'APPLICATION_APPROVED'
  | 'APPLICATION_REJECTED'
  | 'PARTICIPATION_REQUESTED'
  | 'PARTICIPATION_APPROVED'
  | 'PARTICIPATION_REJECTED'
  | 'PARTICIPATION_CANCELED'
  | 'TICKET_ISSUED'
  | 'TICKET_CHECKED_IN'
  | 'CHANNEL_BANNED'
  | 'CHANNEL_UNBANNED'
  | 'REFUND_COMPLETED'
  | 'EVENT_ANNOUNCEMENT'

/**
 * PR95 — 사용자별 NotificationType 수신 선호. backend 응답은 모든 type 을 반환하며 row 가 없는
 * type 은 enabled=true 로 채워진 상태.
 *
 * PR104 — `updatedAt` 은 backend row.updatedAt 의 lightweight signal. row 가 없는 type
 * (사용자가 한 번도 설정한 적 없음 = 기본값) 은 null. 본 필드는 변경 이력을 대체하지 않는다.
 */
export interface NotificationPreference {
  type: NotificationType
  enabled: boolean
  updatedAt?: string | null
}

/** PR95 — 부분 갱신 요청. request 에 없는 type 은 backend 가 기존 값을 유지한다. */
export interface UpdateNotificationPreferencesRequest {
  preferences: Array<{ type: NotificationType; enabled: boolean }>
}
